### ISSUE TITLE
wave.open does not accept PathLike objects

### ISSUE THREAD

**Issue by f8c46d9f-f92e-420f-956b-d795f5ff26ab** [python/cpython#75593]
BPO | [31412](https://bugs.python.org/issue31412)
--- | :---
Nosy | @berkerpeksag, @serhiy-storchaka, @mscuthbert
PRs | <li>python/cpython#3484</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-09-10.18:41:03.935>
labels = ['3.7', 'type-feature', 'library']
title = 'wave.open does not accept PathLike objects'
updated_at = <Date 2017-11-01.09:01:18.609>
user = 'https://github.com/mscuthbert'
```

bugs.python.org fields:
```python
activity = <Date 2017-11-01.09:01:18.609>
actor = 'serhiy.storchaka'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2017-09-10.18:41:03.935>
creator = 'mscuthbert'
dependencies = []
files = []
hgrepos = []
issue_num = 31412
keywords = ['patch']
message_count = 2.0
messages = ['301826', '305361']
nosy_count = 3.0
nosy_names = ['berker.peksag', 'serhiy.storchaka', 'mscuthbert']
pr_nums = ['3484']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue31412'
versions = ['Python 3.7']
```

</p></details>

### Linked PRs
* gh-140951

**Comment by f8c46d9f-f92e-420f-956b-d795f5ff26ab** [comment]
The wave library (reading and writing .wav audio files) accepts filehandles, strings, and bytes, but does not accept PathLike objects.  Patch will modify wave.py to allow open to use these objects.

**Comment by serhiy-storchaka** [comment]
wave is a one of three audio file modules (aifc, sunau, wave) that should have the same interface.

**Comment by hugovk** [comment]
Note the aifc and sunau modules are deprecated in 3.11 and set for removal in 3.13.

See [PEP 594 – Removing dead batteries from the standard library](https://peps.python.org/pep-0594/) and https://github.com/python/cpython/issues/91217.

The wave module is being kept.

**Comment by mscuthbert** [comment]
The code I contributed to 3.7 is still, I believe, the best path forward.  Based on my experiences trying to contribute to cpython, I will not be continuing on this issue.  It can be closed.

**Comment by arhadthedev** [comment]
> Based on my experiences trying to contribute to cpython, I will not be continuing on this issue. It can be closed.

After the migration, the issue pile finally started to actively reduce in size torwards something manageable:

> 
![image](https://user-images.githubusercontent.com/4881073/172370783-2efd544b-63cc-4b15-8be3-016c9357638f.png)

So your issue being accidentally pulled out-of-order from the abyss of 6,878 gives a great chance to finally see it merged.

... after a month, give-or-take, to allow anybody concerned a chance to give a fresh feedback. The important thing now is not to allow it to sink back.

*I suspect that hugovk just noted that in September, 2023 two affected modules will cease to exist. However, we need to live till that date first, so your fix is still relevant in full volume.*

**Comment by AA-Turner** [comment]
The linked PR (#3484) was closed, although it doesn't say by whom -- I think it could be reopened and updated to drop the changes to the deprecated modules.

A

**Comment by arhadthedev** [comment]
> I think it could be reopened and updated to drop the changes to the deprecated modules.

Can we keep them to make backporting easier (the PR fixes divergence from [PEP 519](https://peps.python.org/pep-0519/) so I believe it's a bugfix and should be backported)?

**Comment by AA-Turner** [comment]
@arhadthedev Can you point out the section of PEP-519 that the modules currently diverge from? I think there's a case to be made both ways for bug/feature, so I'm taking the safer option for now and marking this as a feature and not for backport.

If there's agreement it is a bug then we should update this issue and the PR to reflect that.

A

**Comment by arhadthedev** [comment]
> It is expected that most APIs in Python’s standard library that currently accept a file system path will be updated appropriately to accept path objects (whether that requires code or simply an update to documentation will vary)

However, I agree that "expected" and "most" are keywords here probably caused by pitfalls that originally prevented the three modules from this feature.

**Comment by AA-Turner** [comment]
Marking as pending per https://github.com/python/cpython/pull/3484#issuecomment-1151639655, GitHub hasn't been playing nicely with @mscuthbert unfortunatley.

A

**Comment by arhadthedev** [comment]
Superseded by more general gh-102425.

**Comment by Bomme** [comment]
@arhadthedev @AA-Turner  could this issue be reopened after the superseding issues gh-102425 and gh-103659 seem to have been closed or stopped?
I believe that the issue is still valid and it would be nice to (finally) use the wave module with PathLike objects. 😃

**Comment by serhiy-storchaka** [comment]
I agree that the original PR was just buried under the pile of others.

The issue is still valid, although it has low priority. If someone wants to finish it, you are welcome. But if it remains inactive again, it will be closed dues to the lack of interest.

**Comment by dr-carlos** [comment]
Not sure if this is a separate issue, but surely `wave.open()` should also accept bytes objects (as it is just passing the filename to `builtins.open()`)?
